contracts/cw-ics721-bridge/src/contract.rs

```diff
@@ -0,0 +1,539 @@
+            WasmMsg::Instantiate {
+                admin: None, // TODO: Any reason to set ourselves as admin?
+                code_id: CW721_ICS_CODE_ID.load(deps.storage)?,
+                msg: to_binary(
+                    &(cw721_base::msg::InstantiateMsg {
+                        // Name of the collection MUST be class_id as this is how
+                        // we create a map entry on reply.
+                        name: class_id.clone(),
+                        symbol: class_id.clone(), // TODO: What should we put here?
+                        minter: env.contract.address.to_string(),
+                    }),
+                )?,
+                funds: vec![],
+                label: format!("{} ICS721 cw721 backing contract", class_id),
+            },
+            INSTANTIATE_CW721_REPLY_ID,
+        );
+        vec![message]
+    };
+
+    // Mint the requested tokens. Submessages and their replies are
+    // always executed before regular messages [1], so we can sleep
+    // nicely knowing this won't happen until the cw721 contract has
+    // been instantiated.
+    //
+    // [1] https://github.com/CosmWasm/cosmwasm/blob/main/SEMANTICS.md#order-and-rollback
+    let mint_message = WasmMsg::Execute {
+        contract_addr: env.contract.address.into_string(),
+        msg: to_binary(&ExecuteMsg::Mint {
+            class_id,
+            token_ids,
+            token_uris,
+            receiver,
+        })?,
+        funds: vec![],
+    };
+
+    Ok(Response::default()
+        .add_attribute("method", "do_instantiate_and_mint")
+        .add_submessages(submessages)
+        .add_message(mint_message))
+}
+
+fn execute_receive_nft(
+    deps: DepsMut,
+    info: MessageInfo,
+    token_id: String,
+    sender: String,
+    msg: Binary,
+) -> Result<Response, ContractError> {
+    let sender = deps.api.addr_validate(&sender)?;
+    let msg: IbcAwayMsg = from_binary(&msg)?;
+
+    // When we receive a NFT over IBC we want to avoid parsing of the
+    // classId field at all costs. String parsing has interesting time
+    // complexity issues and is error prone.
+    //
+    // We would normally need to parse the classId field when we
+    // receive a NFT from another chain, and the top of the classId
+    // stack is our own chain. In this case, it is possible that upon
+    // that NFT arriving at this chain it will be unpacked back into a
+    // native NFT. To determine this we could either:
+    //
+    // (1) Check if the classId has no '/' characters. Or,
+    // (2) Store each NFT we receive in our maps (giving it a classID
+    //     of its own address) and do the same lookup / transfer
+    //     regardless of if the NFT is becoming a native one.
+    //
+    // To avoid special logic when receiving packets, we go the save
+    // route. We can change this back later if we'd like.
+    if !NFT_CONTRACT_TO_CLASS_ID.has(deps.storage, info.sender.clone()) {
```

Comment:
Can we clarify this comment a bit? It's still a little confusing IMO... we have two double maps to the same address?

Reply:
Maybe add a note that the actually class ID is added in a later step?

Reply:
Clarifying here in the new diagram design that we do in fact need two maps. One for forwards and one for backwards.

CLASS_ID_TO_NFT_CONTRACT => This is how you know which contract corresponds to which class id
NFT_CONTRACT_TO_CLASS_ID => When you are sending outbound, the bridge won't be able to tell the difference between any given Cw721 contract since "escrow" contracts are now collection contracts (the same as any other contract). class_id is not stored in standard Cw721 contracts, therefore on send out, you need to look up from the contract address whether you have stored a class id for this before or not.